br/pkg/stream/stream_metas.go

```diff
@@ -408,3 +321,956 @@ func ReplaceMetadata(meta *backuppb.Metadata, filegroups []*backuppb.DataFileGro
+	// NOTE: Execution of truncation wasn't implemented here.
+	// If we are going to truncate some files, for now we still need to use `br log truncate`.
+	for _, compaction := range mig.Compactions {
+		// Can we also remove the compaction when `until-ts` is equal to `truncated-to`...?
+		if compaction.CompactionUntilTs > mig.TruncatedTo {
+			result.NewBase.Compactions = append(result.NewBase.Compactions, compaction)
+		} else {
+			m.tryRemovePrefix(ctx, compaction.Artifacts, result)
+			m.tryRemovePrefix(ctx, compaction.GeneratedFiles, result)
+		}
+	}
+	for _, pfx := range mig.DestructPrefix {
+		m.tryRemovePrefix(ctx, pfx, result)
+	}
+
+	result.NewBase.TruncatedTo = mig.TruncatedTo
+
+	m.Hooks.StartLoadingMetaForTruncating()
+	mdSet := new(StreamMetadataSet)
+	mdSet.MetadataDownloadBatchSize = 128
+	shiftTS, err := mdSet.LoadUntilAndCalculateShiftTS(ctx, m.s, mig.TruncatedTo)
+	if err != nil {
+		result.Warnings = append(result.Warnings, errors.Annotatef(err, "failed to open meta storage"))
+		return
+	}
+	m.Hooks.EndLoadingMetaForTruncating()
+
+	m.doTruncateLogs(ctx, mdSet, shiftTS, result)
+}
+
+func (m MigrationExt) loadFilesOfPrefix(ctx context.Context, prefix string) (out []string, err error) {
+	err = m.s.WalkDir(ctx, &storage.WalkOption{SubDir: prefix}, func(path string, size int64) error {
+		out = append(out, path)
+		return nil
+	})
+	return
+}
+
+// doTruncateLogs truncates the logs until the specified TS.
+// This might be slow.
+func (m MigrationExt) doTruncateLogs(
+	ctx context.Context,
+	metadataInfos *StreamMetadataSet,
+	from uint64,
+	out *MigratedTo,
+) {
+	mu := new(sync.Mutex)
+	updateResult := func(f func(r *MigratedTo)) {
+		mu.Lock()
+		defer mu.Unlock()
+
+		f(out)
+	}
+	emitErr := func(err error) {
+		updateResult(func(r *MigratedTo) {
+			r.Warnings = append(r.Warnings, err)
+		})
+	}
+	cannotBeRetryByRerunBase := func(err error) error {
+		return errors.Annotate(
+			err,
+			"this error may not be retry by `migrate-to --base`, "+
+				"you may need to rerun `log truncate`",
+		)
+	}
+
+	worker := util.NewWorkerPool(128, "delete files")
+	wg := new(sync.WaitGroup)
+
+	m.Hooks.StartDeletingForTruncating(metadataInfos, from)
+	defer m.Hooks.DeletedAllFilesForTruncating()
+	for path, metaInfo := range metadataInfos.metadataInfos {
+		path := path
+		if metaInfo.MinTS >= from {
+			continue
+		}
+		wg.Add(1)
+		worker.Apply(func() {
+			defer wg.Done()
+			data, err := m.s.ReadFile(ctx, path)
+			if err != nil {
+				emitErr(cannotBeRetryByRerunBase(
+					errors.Annotatef(err, "failed to open meta %s", path)))
+				return
+			}
+
+			// Note: maybe make this a static method or just a normal function...
+			meta, err := (*MetadataHelper).ParseToMetadataHard(nil, data)
+			if err != nil {
+				emitErr(cannotBeRetryByRerunBase(
+					errors.Annotatef(err, "failed to parse meta %s", path)))
+				return
+			}
+
+			me := new(pb.MetaEdit)
+			me.Path = path
+			for _, ds := range meta.FileGroups {
```

Comment:
should we also append the `me.DeleteLogicalFiles` because `m.applyMetaEditTo` uses it?